### PR TITLE
Add new publish with job id method to keep jobid consistent for secondary storage

### PR DESCRIPTION
**Pull request by Napoleon0621**
#168 
1. add PublishWithJobID method to engine in order to keep  job id consistent when secondary storage is enabled.
2. add related unit tests.

### CODE REVIEW

File: engine/migration/engine_test.go [hunk 1]
```diff
@@ -227,3 +227,20 @@ func TestEngine_DeadLetter_Size(t *testing.T) {
...
+	jobID, err := e.PublishWithJobID("ns-engine", "q10", "jobID1",
+		body, 10, 0, 1)
+	t.Log(jobID)
+	if err != nil {
+		t.Fatalf("Failed to publish: %s", err)
+	}
+	// Make sure the new engine received the job
+	job, err := NewRedisEngine.Consume("ns-engine", []string{"q10"}, 3, 0)
+	if job.ID() != jobID {
+		t.Fatalf("NewRedisEngine should received the job:%s, but get: %s", jobID, job.ID())
```
Comment (iwinux) [inline engine/migration/engine_test.go:244]: Use `github.com/stretchr/testify` for assertions?

**Review by git-hulk — COMMENTED** [review]
I think we can use `PublishWithJobID` to reimplement `Publish` to avoid duplicate code.

File: engine/engine.go [hunk 2]
```diff
@@ -22,4 +22,5 @@ type Engine interface {
 	Shutdown()
 
 	DumpInfo(output io.Writer) error
+	PublishWithJobID(namespace, queue, storedJobID string, body []byte, ttlSecond, delaySecond uint32, tries uint16) (jobID string, err error)
```
Comment (iwinux) [inline engine/engine.go:25]: Maybe `PublishJob(req JobRequest) (jobID string, err error)`?

File: engine/migration/engine_test.go (hunk 1 above)
Comment (git-hulk) [reply engine/migration/engine_test.go:244]: Agree, we should test suites to testify instead of using legacy way.

File: engine/engine.go (hunk 2 above)
Comment (Napoleon0621) [reply engine/engine.go:25]: > Maybe `PublishJob(req JobRequest) (jobID string, err error)`?

yes, will change the API schema.

[the remark this answers is not in the document]
Comment (Napoleon0621) [reply engine/engine.go:25]: Will extract the common code into a new function.

Comment (git-hulk) [reply engine/engine.go:25]: I think this API should be `PublishJob(job engine.Job) (jobID string, error)`

File: engine/redis/engine.go [hunk 3]
```diff
@@ -75,28 +76,7 @@ func (e *Engine) Publish(namespace, queue string, body []byte, ttlSecond, delayS
 	e.meta.RecordIfNotExist(namespace, queue)
 	e.monitor.MonitorIfNotExist(namespace, queue)
 	job := engine.NewJob(namespace, queue, body, ttlSecond, delaySecond, tries)
```
Comment (git-hulk) [inline engine/redis/engine.go:78]: should remove above codes since they are done in `PublishJob`

File: engine/redis/engine.go [hunk 4]
```diff
@@ -282,3 +262,43 @@ func (e *Engine) DumpInfo(out io.Writer) error {
...
+	if storedJobID == "" {
+		return "", errors.New("PublishWithJobID failed: null job id")
+	}
+	e.meta.RecordIfNotExist(namespace, queue)
+	e.monitor.MonitorIfNotExist(namespace, queue)
+	job := engine.NewJobWithJobID(namespace, queue, body, ttlSecond, delaySecond, tries, storedJobID)
+	return e.PublishJobs(job)
+}
+
+func (e *Engine) PublishJobs(job engine.Job) (jobID string, err error) {
```
Comment (git-hulk) [inline engine/redis/engine.go:282]: PublishJob NOT PublishJobs

**Comment by Napoleon0621** [comment]
@git-hulk plz take another look, I refactor the Publish API schema. Thx.

**Comment by git-hulk** [comment]
Current implementation looks good to me except I think we should minimize the modification of the old Redis engine. So I advice that we can only unpack variables from job and don't do other modification. Like below:
```
Publish(job Job) error {
 namespace := job.Namespace()
 queue := job.Queue
 ...
}
```

**Comment by Napoleon0621** [comment]
>  
Updates made here, please help review the change at your convenience, thx.

**Review by git-hulk — APPROVED** [review]

**Pull request merged**